addons/mail/models/mail_activity.py

```diff
@@ -378,6 +378,95 @@ def _check_access_assignation(self):
+
+        self.flush(['res_model', 'res_id'])
+        activities_to_check = []
+        for sub_ids in self._cr.split_for_in_conditions(ids):
+            self._cr.execute("""
+                SELECT DISTINCT a.id, a.res_model, a.res_id
+                FROM "%s" a
+                WHERE a.id = ANY (%%(ids)s)""" % self._table, dict(ids=list(sub_ids)))
+            activities_to_check = self._cr.dictfetchall()
+
+        activity_to_documents = {}
+        for activity in activities_to_check:
+            activity_to_documents.setdefault(activity['res_model'], list()).append(activity['res_id'])
+
+        allowed_ids = []
+        for doc_model, doc_ids in activity_to_documents.items():
+            # fall back on related document access right checks. Use the same as defined for mail.thread
+            # if available; otherwise fall back on read
+            if hasattr(self.env[doc_model], '_mail_post_access'):
+                doc_operation = self.env[doc_model]._mail_post_access
+            else:
+                doc_operation = 'read'
+            right = self.env[doc_model].check_access_rights(doc_operation, raise_exception=False)
+            if right:
+                valid_docs = self.env[doc_model].browse(doc_ids)._filter_access_rules(doc_operation)
+                allowed_ids += [
+                    activity['id'] for activity in activities_to_check
+                    if activity['res_model'] == doc_model and activity['res_id'] in valid_docs.ids]
+
+        if count:
+            return len(allowed_ids)
+        else:
+            # re-construct a list based on ids, because set did not keep the original order
+            id_list = [id for id in ids if id in allowed_ids]
+            return id_list
+
+    @api.model
+    def _read_group_raw(self, domain, fields, groupby, offset=0, limit=None, orderby=False, lazy=True):
+        """ Override that adds specific access rights of mail.activity, to remove
+        ids uid could not see according to our custom rules. Please refer to
+        _filter_access_rules_remaining for more details about those rules.
+
+        Here, we simply re-use the _search implementation to force the read group
+        domain to allowed ids only. """
+
+        # Rules do not apply to administrator
+        if not self.env.is_superuser():
+            allowed_ids = self._search(domain, count=False)
+            if allowed_ids:
+                domain = expression.AND([domain, [('id', 'in', allowed_ids)]])
+            else:
+                # force void result if no allowed ids found
+                domain = expression.AND([domain, [(0, '=', 1)]])
+
+        return super(MailActivity, self)._read_group_raw(
+            domain=domain, fields=fields, groupby=groupby, offset=offset,
+            limit=limit, orderby=orderby, lazy=lazy,
+        )
+
+    def read(self, fields=None, load='_classic_read'):
+        """ Override to explicitely call check_access_rule, that is not called
```

Comment:
If we want to be more complete, taken from task (maybe to rephrase): "When reading, read calls _read that manually applies ir rules (_apply_ir_rules) if there are fields to read, otherwise calls check_access_rule('read') (strange behavior). So indeed _filter_access_rules and _filter_access_rules_python are not called."